File: 04. Best Time to Buy and Sell Stock/left-right iterators.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int left = prices[0], right = prices[0], profit = 0, max_profit = 0;

        for(int price: prices)
        {
            if(price < left)
                {
                left = price;
                right = price;
                }
            else if(right < price)
            {
                right = price;
                profit = right - left;
            }
            if(profit > max_profit)
                max_profit = profit;
        }

        return max_profit;
    }
};

File: 04. Best Time to Buy and Sell Stock/left-right iterators.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        left, right = prices[0], prices[0]
        profit, max_profit = 0, 0

        for price in prices:
            if price < left:
                left = price
                right = price
            elif price > right:
                right = price
                profit = right - left

                if profit > max_profit:
                    max_profit = profit
        
        return max_profit